### ISSUE TITLE
Support null-safety

### ISSUE THREAD

**Issue by mehmetf** [tekartik/sqflite#529]
Opening a tracking issue to migrate this repo to use null-safe code. Let's discuss how we can arrange this. The best way is to create a branch that is kept in sync with `master` until the code can be pushed to master.

@alextekartik @davidmorgan as FYI

**Comment by alextekartik** [comment]
My original plan was to have the `synchronized` package publish first before tackling the migration of `sqflite` (and also `sqflite_common`)  since I think it is better to have all its dependencies migrated first. And yes I was thinking following the migration process proposed (creating a null safety branch) but I want to avoid maintaining 2 branches for too long.

**Comment by hacker1024** [comment]
At the moment, this package won't even compile with NNBD enabled. Several errors like this are given:

```
Pub/Cache/hosted/pub.dartlang.org/sqflite-1.3.2+1/lib/src/compat.dart:1:1: Error: Null safe libraries are not allowed to export declarations from of opt-out libraries.
export 'package:sqflite_common/src/compat.dart';
```

**Comment by alextekartik** [comment]
The work is in progress in the `null_safety` branch: https://github.com/tekartik/sqflite/pull/535

**Comment by alextekartik** [comment]
I have a doubt currently regarding keeping `Map<String, dynamic>` and `List<dynamic>` to replace them with `Map<String, Object?>` and `List<Object?>` or `List<Object>` since sometimes null are not allowed.

@mehmetf (or others) Do you have any recommendation on this? I'd like to take the opportunity of the breaking nnbd change if anything is needed.

**Comment by mehmetf** [comment]
I am with you. I have gone from `dynamic` to `Object[?]` in many places during my own migrations simply because dynamic does not encode nullability.

Many Dart coders use `dynamic` instead of Object because it feels more readable. However, the true difference (i.e. dynamic _dispatch_)  is rarely utilized. Dart style guide also recommends against it: https://dart.dev/guides/language/effective-dart/design#avoid-using-dynamic-unless-you-want-to-disable-static-checking

**Comment by alextekartik** [comment]
Thanks @mehmetf I did make the change and I'm quite happy with it. It allows to remove null unit tests that were just failing anyway. I published a dev version.

**Comment by simolus3** [comment]
For moor, it would be much easier if `arguments` was a `List<Object?>` since it allows binding null to variables. I think this can be useful outside of moor too: sqlite3 has `sqlite3_bind_null` and sqflite's current Android implementation also supports null parameters. I'm sure that there are sqflite users out there passing null in an argument list, and that has always been valid. Can this still be changed?

**Comment by alextekartik** [comment]
I switch from `List<dynamic>` to `List<Object>` (after seeing what you have done on sqlite3 orginally @simolus3 - but you seem to have change that) on purpose because I could not find an example of binding a null value. For query we have to use `IS NULL` instead if `= ?` (see my discussion here: https://github.com/tekartik/sqflite/pull/535#pullrequestreview-536015163).

`Map<String, Object?>` is however used for insert/update helper (but here sqflite replace the null value inline).

Do you have an example where this could be useful? (and work cross platform)

**Comment by simolus3** [comment]
Oops, yeah I meant to use `List<Object?>` but it seems like I forgot one case that I've now fixed. Moor also replaces `null` variables with their literal, so it should be fine for me. 

> `Map<String, Object?>` is however used for insert/update helper (but here sqflite replace the null value inline).

That sounds reasonable, but users might want to use `null` in more complex inserts or updates (e.g. using common table expressions or something similar), so it might be reasonable for `rawInsert` and `rawUpdate`. There are valid uses of `NULL` in queries too. For instance, if we had something like `SELECT CASE some_condition THEN name ELSE ? END FROM users` to select names or a default value provided in Dart. It's perfectly valid to bind `?` to `null` here.

I don't know if all platforms allow binding variables to `null`. If some don't, I can live with non-nullable parameters too. But arbitrarily disallowing `null` elements sounds like an unnecessary limitation to me.

**Comment by alextekartik** [comment]
Ok I will follow you advice.

**Comment by alextekartik** [comment]
@simolus3 So I did some testing:

```dart
await db.rawQuery('SELECT CASE WHEN 0 = 1 THEN 1 ELSE ? END', [null]);
```

It works using ffi (your sqlite3 package) but fails on Android (using sqflite, so it seems the native android API does not allow it unlike you mentioned - `DatabaseException(the bind value at index 1 is null)`). I have yet to test on iOS. So I remain puzzled on this. Having `List<Object>` was good to avoid such mistake. However I feel enclined to switch to `List<Object?>` like you propose to avoid this being a limitation in the future or for people only using ffi.

**Comment by alextekartik** [comment]
@simolus3 I published a new version of everything (sqlite/common/ffi) supporting now `List<Object?>` everywhere.

**Comment by simolus3** [comment]
Thank you! And sorry for causing confusion about nulls on Android. I assumed it would work based on the `null` checks in `SqlCommand.toString()`, but those seem to be unrelated.

**Comment by alextekartik** [comment]
The worst of the android native API is that every parameter are String...too bad they never changed that.
```
public Cursor rawQuery(String sql, String[] selectionArgs)
```

**Comment by asartalo** [comment]
I'm currently using sqflite3 and on builds, I usually get these errors:

```
ERROR: <TRIMMED>/sqlite3-0.1.9-nullsafety.2/lib/src/ffi/sqlite3.ffi.dart:6:7: Info: Struct 'char' is empty. Support for empty structs is deprecated and will be removed in the next stable version of Dart. Use Opaque instead.
ERROR: class char extends Struct {}
```
This has been fixed in v1.0.0 of sqlite3 which is null-safe already. Would it be possible to have sqflite_common_ffi use it instead of `>=0.1.9-nullsafety.0 <1.0.0`?

**Comment by simolus3** [comment]
The library migrated away from empty structs in `0.1.10-nullsafety.0`, and there aren't any breaking changes between that and `1.0.0`. 
So you could use `>=0.1.10-nullsafety.0 <2.0.0` as a safe constraint with maximum compatibility (or even use `0.1.9` as lower bound, but then users would have to `pub upgrade` to avoid warnings).

**Comment by alextekartik** [comment]
Thanks @asartalo . Just published `sqflite_common_ffi 2.0.0-nullsafety.7`. If pub publish would not complain I would remove any upper dependency (`>=0.1.10-nullsafety.0` - how can we predict it will broken?) and only enforce them when something breaks.

**Comment by asartalo** [comment]
@alextekartik thank you! I'm using it now and it works.

**Comment by Rockvole** [comment]
Is null safety supposed to work with version 2.0.0 ? From a basic project (flutter create). I add :
pubspec.yaml
```
environment:
  sdk: ">=2.12.0-0 <3.0.0"

dependencies:
  sqflite: ^2.0.0
```

On pub get, I see :
```
Because every version of integration_test from sdk depends on path 1.8.0-nullsafety.3 and sqflite >=2.0.0 depends on path >=1.8.0 <3.0.0, integration_test from sdk is incompatible with sqflite >=2.0.0.
So, because rockvole_replicator_todo depends on both sqflite ^2.0.0 and integration_test any from sdk, version solving failed.
pub get failed (1; So, because rockvole_replicator_todo depends on both sqflite ^2.0.0 and integration_test any from sdk, version solving failed.)
```

**Comment by alextekartik** [comment]
@Rockvole I could not reproduce. That's a weird fixed dependency though. I published sqflite 2.0.0+1 that supports path 1.8.0-0 that should fix your issue. Are you on the latest beta sdk?

**Comment by Rockvole** [comment]
@alextekartik - thanks that fixed it. I am on :
```
Flutter 1.27.0-9.0.pre.172 • channel master • https://github.com/flutter/flutter.git
Framework • revision d3209379ba (17 hours ago) • 2021-03-02 16:34:05 -0800
Engine • revision cd0e54348b
Tools • Dart 2.13.0 (build 2.13.0-93.0.dev)
```
